### ISSUE TITLE
Regression in ResultQuery.fetchMap(Field, Field) with null values

### ISSUE THREAD

**Issue by arlampin** [jOOQ/jOOQ#12123]
### Expected behavior

ResultQuery.fetchMap(Field, Field) should not throw NPE when value field maps to a null value. This was the behavior on 3.14.x. fetchMap's docs don't list any limitations on value field, so I assume this is unintentional. Probably also applies to other fetchMap overrides.

### Actual behavior

NPE is thrown internally from HashMap.merge, at least with Java 16.

```
java.lang.NullPointerException
	at java.base/java.util.HashMap.merge(HashMap.java:1359)
	at java.base/java.util.stream.Collectors.lambda$toMap$68(Collectors.java:1666)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.jooq.impl.AbstractCursor.collect(AbstractCursor.java:78)
	at org.jooq.impl.ResultQueryTrait.collect(ResultQueryTrait.java:358)
	at org.jooq.impl.ResultQueryTrait.fetchMap(ResultQueryTrait.java:850)
```

### Steps to reproduce the problem

An example query, where TEAM.BUSINESS_ID is a nullable column.

```java
        var businessIds = dsl.select()
            .from(TEAM)
            .fetchMap(TEAM.ID, TEAM.BUSINESS_ID);
```

### Versions

- jOOQ: 3.15.0
- Java: 16

**Comment by lukaseder** [comment]
Thanks a lot for your report! Will look into this shortly

**Comment by lukaseder** [comment]
`Result.intoGroups()` is also affected when there are `null` keys. There is no reason to disallow that.

**Comment by lukaseder** [comment]
We can't use `Collectors.toMap()` with the merge operation, see: https://stackoverflow.com/q/24630963/521799, https://bugs.openjdk.java.net/browse/JDK-8148463

I'll roll my own `Collector` instead.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.1 (#12124). Thanks again for your report!
